### PR TITLE
[ALLUXIO-2020] Adding cp command for the Alluxio shell.

### PR DESCRIPTION
**Pull request by jsimsa**
Addresses https://alluxio.atlassian.net/browse/ALLUXIO-2020

### CODE REVIEW

File: shell/src/main/java/alluxio/shell/command/CpCommand.java [hunk 1]
```diff
@@ -0,0 +1,236 @@
...
+      FileOutStream os = closer.register(mFileSystem.createFile(dstPath, createFileOptions));
+      IOUtils.copy(is, os);
+      System.out.println("Copied " + srcPath + " to " + dstPath.getPath());
+    } catch (AlluxioException e) {
+      throw new IOException(e.getMessage());
+    }
+  }
+
+  @Override
+  public String getUsage() {
```
Comment (ifcharming) [inline shell/src/main/java/alluxio/shell/command/CpCommand.java]: Do you think we should add `-R` option for copy dir ? to make it more similar as linux `cp` semantics.

File: tests/src/test/java/alluxio/shell/command/CpCommandTest.java [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+import org.apache.commons.io.IOUtils;
+import org.junit.Assert;
+import org.junit.Test;
+
+import java.io.IOException;
+
+/**
+ * Tests for cp command.
+ */
+public class CpCommandTest extends AbstractAlluxioShellTest {
```
Comment (ifcharming) [inline tests/src/test/java/alluxio/shell/command/CpCommandTest.java]: nit: final

File: tests/src/test/java/alluxio/shell/command/CpCommandTest.java [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+  @Test
+  public void copyDirTest() throws Exception {
```
Comment (ifcharming) [inline tests/src/test/java/alluxio/shell/command/CpCommandTest.java]: nit: a brief description for the test case?

File: tests/src/test/java/alluxio/shell/command/CpCommandTest.java [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+    Assert.assertTrue(
+        equals(new AlluxioURI("/copy/foo/foobar1"), new AlluxioURI("/testWildCards/foo/foobar1")));
+    Assert.assertTrue(
+        equals(new AlluxioURI("/copy/foo/foobar2"), new AlluxioURI("/testWildCards/foo/foobar2")));
+    Assert.assertTrue(
+        equals(new AlluxioURI("/copy/foobar4"), new AlluxioURI("/testWildCards/foobar4")));
+  }
+
+  @Test
+  public void copyWildcardDirTest() throws Exception {
```
Comment (ifcharming) [inline tests/src/test/java/alluxio/shell/command/CpCommandTest.java]: nit: a brief description for the test case?

**Comment by ifcharming** [comment]
Thanks @jsimsa for adding the new command. I left a few comments.

File: shell/src/main/java/alluxio/shell/command/CpCommand.java (hunk 1 above)
Comment (jsimsa) [inline shell/src/main/java/alluxio/shell/command/CpCommand.java]: if source is a directory, it will be copied recursively

File: tests/src/test/java/alluxio/shell/command/CpCommandTest.java (hunk 2 above)
Comment (jsimsa) [inline tests/src/test/java/alluxio/shell/command/CpCommandTest.java]: done

File: tests/src/test/java/alluxio/shell/command/CpCommandTest.java (hunk 3 above)
Comment (jsimsa) [inline tests/src/test/java/alluxio/shell/command/CpCommandTest.java]: done

File: tests/src/test/java/alluxio/shell/command/CpCommandTest.java (hunk 4 above)
Comment (jsimsa) [inline tests/src/test/java/alluxio/shell/command/CpCommandTest.java]: done

**Comment by jsimsa** [comment]
@ifcharming addressed your comments, PTAL

File: shell/src/main/java/alluxio/shell/command/CpCommand.java [hunk 5]
```diff
@@ -0,0 +1,236 @@
...
+    }
+  }
+
+  /**
+   * Copies a list of files or directories specified by srcPaths to the destination specified by
+   * dstPath. This method is used when the original source path contains wildcards.
+   *
+   * @param srcPaths a list of files or directories in the Alluxio filesystem
+   * @param dstPath the destination in the Alluxio filesystem
+   * @throws IOException
```
Comment (aaudiber) [inline shell/src/main/java/alluxio/shell/command/CpCommand.java]: Would it make sense to `Throwables.propagate` exceptions thrown in this class instead of wrapping everything in `IOException`?

File: shell/src/main/java/alluxio/shell/command/CpCommand.java [hunk 6]
```diff
@@ -0,0 +1,236 @@
...
+   */
+  private void copyWildcard(List<AlluxioURI> srcPaths, AlluxioURI dstPath) throws IOException {
```
Comment (aaudiber) [inline shell/src/main/java/alluxio/shell/command/CpCommand.java]: This method doesn't actually deal with wildcards, maybe call it copyPaths?

Comment (aaudiber) [inline shell/src/main/java/alluxio/shell/command/CpCommand.java]: nvm, the error message refers to them

File: shell/src/main/java/alluxio/shell/command/CpCommand.java (hunk 5 above)
Comment (jsimsa) [inline shell/src/main/java/alluxio/shell/command/CpCommand.java]: What would the benefit of doing that? The reason for wrapping everything in `IOException` are:

a) I am following the precedent set for other command classes
b) the AlluxioShell wrapper around `run` actually catches IOException and prints it to `LOG.error` if we use `Throwables.propagate` then we lose that functionality

Comment (aaudiber) [inline shell/src/main/java/alluxio/shell/command/CpCommand.java]: The benefits are that (1) we won't have to throw non-IO related exceptions as IOExceptions and (2) we don't need exception boilerplate. It seems like the wrapper ought to log any exception that's thrown, not just IOExceptions.

Since the precedent has already been set in other classes we can leave changing the use of IOException for a future PR, but if we keep `@throws IOException` we should include a nonempty javadoc message.

File: tests/src/test/java/alluxio/shell/command/CpCommandTest.java [hunk 7]
```diff
@@ -0,0 +1,132 @@
...
+  /**
+   * Tests copying a list of files specified through a wildcard expression.
+   */
+  @Test
+  public void copyWildcardDirTest() throws Exception {
+    AlluxioShellUtilsTest.resetFileHierarchy(mFileSystem);
+    int ret = mFsShell.run("cp", "/testWildCards/*/foo*", "/copy");
+    Assert.assertEquals(0, ret);
+    Assert.assertTrue(mFileSystem.exists(new AlluxioURI("/copy")));
+    Assert.assertTrue(mFileSystem.exists(new AlluxioURI("/copy/foobar1")));
```
Comment (uronce-cc) [inline tests/src/test/java/alluxio/shell/command/CpCommandTest.java]: We delibrately do not preserve the structure of sub-directories?

File: tests/src/test/java/alluxio/shell/command/CpCommandTest.java [hunk 8]
```diff
@@ -0,0 +1,132 @@
...
+  /**
+   * Tests invalid input arguments.
+   */
+  @Test
+  public void copyWildcardNotDirTest() throws Exception {
+    AlluxioShellUtilsTest.resetFileHierarchy(mFileSystem);
+    int ret;
+    ret = mFsShell.run("cp", "/testWildCards", "/testWildCards/foobar4");
+    Assert.assertEquals(-1, ret);
+    ret = mFsShell.run("cp", "/testWildCards/*", "/testWildCards/foobar4");
```
Comment (uronce-cc) [inline tests/src/test/java/alluxio/shell/command/CpCommandTest.java]: Curious about the behavior of "cp /testWildCards /testWildCards".

File: tests/src/test/java/alluxio/shell/command/CpCommandTest.java [hunk 9]
```diff
@@ -0,0 +1,132 @@
...
+
+    if (status1.isPersisted() != status2.isPersisted()) {
+      return false;
+    }
+
+    try (Closer closer = Closer.create()) {
+      OpenFileOptions openFileOptions = OpenFileOptions.defaults().setReadType(ReadType.NO_CACHE);
+      FileInStream is1 = closer.register(mFileSystem.openFile(file1, openFileOptions));
+      FileInStream is2 = closer.register(mFileSystem.openFile(file2, openFileOptions));
+      IOUtils.contentEquals(is1, is2);
```
Comment (uronce-cc) [inline tests/src/test/java/alluxio/shell/command/CpCommandTest.java]: `if (!xxx) return false`.

**Comment by uronce-cc** [comment]
@jsimsa Thanks, I left some comment on tests.

File: shell/src/main/java/alluxio/shell/command/CpCommand.java (hunk 1 above)
Comment (apc999) [inline shell/src/main/java/alluxio/shell/command/CpCommand.java]: for `fs rm` command, we require `-R` to delete a dir
```bash
rm: cannot remove a directory, please try rm -R <path>
```
should we keep the commandline style consistent?

File: shell/src/main/java/alluxio/shell/command/CpCommand.java (hunk 5 above)
Comment (apc999) [inline shell/src/main/java/alluxio/shell/command/CpCommand.java]: javadoc exception

File: shell/src/main/java/alluxio/shell/command/CpCommand.java [hunk 10]
```diff
@@ -0,0 +1,236 @@
...
+      srcStatus = mFileSystem.getStatus(srcPath);
+    } catch (AlluxioException e) {
+      throw new IOException(e.getMessage());
+    }
+
+    try (Closer closer = Closer.create()) {
+      OpenFileOptions openFileOptions = OpenFileOptions.defaults().setReadType(ReadType.NO_CACHE);
+      FileInStream is = closer.register(mFileSystem.openFile(srcPath, openFileOptions));
+      CreateFileOptions createFileOptions = CreateFileOptions.defaults();
+      if (srcStatus.isPersisted()) {
```
Comment (apc999) [inline shell/src/main/java/alluxio/shell/command/CpCommand.java]: not sure if I agree the logic here. It seems we are guessing whether to persist dst by the persisted state of src. If this is the case, if you specify -D`writetype` will be overwritten. is this what we want? in other words, should we just let users specify writetype on command line instead of hardcode it here?

File: shell/src/main/java/alluxio/shell/command/CpCommand.java [hunk 11]
```diff
@@ -0,0 +1,236 @@
...
+  private void copyFile(AlluxioURI srcPath, AlluxioURI dstPath) throws IOException {
+    URIStatus srcStatus;
+    try {
+      srcStatus = mFileSystem.getStatus(srcPath);
+    } catch (AlluxioException e) {
+      throw new IOException(e.getMessage());
+    }
+
+    try (Closer closer = Closer.create()) {
+      OpenFileOptions openFileOptions = OpenFileOptions.defaults().setReadType(ReadType.NO_CACHE);
```
Comment (apc999) [inline shell/src/main/java/alluxio/shell/command/CpCommand.java]: why hardcode `NO_CACHE` here (rather than respecting what ppl set in command line using -D or in site properties)?

File: shell/src/main/java/alluxio/shell/command/CpCommand.java [hunk 12]
```diff
@@ -0,0 +1,236 @@
...
+   * dstPath. This method is used when the original source path contains wildcards.
+   *
+   * @param srcPaths a list of files or directories in the Alluxio filesystem
+   * @param dstPath the destination in the Alluxio filesystem
+   * @throws IOException
+   */
+  private void copyWildcard(List<AlluxioURI> srcPaths, AlluxioURI dstPath) throws IOException {
+    URIStatus dstStatus = null;
+    try {
+      dstStatus = mFileSystem.getStatus(dstPath);
```
Comment (apc999) [inline shell/src/main/java/alluxio/shell/command/CpCommand.java]: does `mFileSystem.exists(dstPath)` work here rather than `getFileStatus`

File: shell/src/main/java/alluxio/shell/command/CpCommand.java [hunk 13]
```diff
@@ -0,0 +1,236 @@
...
+      throw new IOException(e.getMessage());
+    }
+
+    if (dstStatus != null && !dstStatus.isFolder()) {
+      throw new IOException(
+          ExceptionMessage.DESTINATION_FILE_CANNOT_EXIST_WITH_WILDCARD_SOURCE.getMessage());
+    }
+    if (dstStatus == null) {
+      try {
+        mFileSystem.createDirectory(dstPath);
```
Comment (apc999) [inline shell/src/main/java/alluxio/shell/command/CpCommand.java]: When `fs cp /src_file /dst_file`, will this create a dir called `/dst_file`?

Comment (apc999) [inline shell/src/main/java/alluxio/shell/command/CpCommand.java]: if we want to match linux cp semantics, we need to check if `dstPath` exists. If not, copy `src` to `parent(dstPath)`; if `dstPath` exists and it is a dir, copy `src` to `dstPath`; if exists and it is a file, throw exception?

File: shell/src/main/java/alluxio/shell/command/CpCommand.java [hunk 14]
```diff
@@ -0,0 +1,236 @@
...
+  }
+
+  /**
+   * Copies a file or a directory in the Alluxio filesystem.
+   *
+   * @param srcPath the source {@link AlluxioURI} (could be a file or a directory)
+   * @param dstPath the destination path in the Alluxio filesystem
+   * @throws IOException
+   */
+  private void copy(AlluxioURI srcPath, AlluxioURI dstPath) throws IOException {
```
Comment (apc999) [inline shell/src/main/java/alluxio/shell/command/CpCommand.java]: possible to reuse code between `copy` and `copyWIldCard`?

File: tests/src/test/java/alluxio/shell/command/CpCommandTest.java [hunk 15]
```diff
@@ -0,0 +1,132 @@
...
+    Assert.assertEquals(0, ret);
+    Assert.assertTrue(mFileSystem.exists(new AlluxioURI("/copy")));
+    Assert.assertTrue(mFileSystem.exists(new AlluxioURI("/copy/bar")));
+    Assert.assertTrue(mFileSystem.exists(new AlluxioURI("/copy/bar/foobar3")));
+    Assert.assertTrue(mFileSystem.exists(new AlluxioURI("/copy/foo")));
+    Assert.assertTrue(mFileSystem.exists(new AlluxioURI("/copy/foo/foobar1")));
+    Assert.assertTrue(mFileSystem.exists(new AlluxioURI("/copy/foo/foobar2")));
+    Assert.assertTrue(mFileSystem.exists(new AlluxioURI("/copy/foobar4")));
+
+    Assert.assertTrue(
```
Comment (apc999) [inline tests/src/test/java/alluxio/shell/command/CpCommandTest.java]: `assertEqual`?

File: tests/src/test/java/alluxio/shell/command/CpCommandTest.java [hunk 16]
```diff
@@ -0,0 +1,132 @@
...
+/**
+ * Tests for cp command.
+ */
+public final class CpCommandTest extends AbstractAlluxioShellTest {
+
+  /**
+   * Tests recursively copying a directory.
+   */
+  @Test
+  public void copyDirTest() throws Exception {
```
Comment (apc999) [inline tests/src/test/java/alluxio/shell/command/CpCommandTest.java]: this tests copy dir when dst doesn't exist.
could you add a test to do copy when (1) dst exists and is a file (expecting a failure), and (2) dst exists and is a dir (expect src to become a subdir in dst)

File: tests/src/test/java/alluxio/shell/command/CpCommandTest.java [hunk 17]
```diff
@@ -0,0 +1,132 @@
...
+  public void copyWildcardNotDirTest() throws Exception {
+    AlluxioShellUtilsTest.resetFileHierarchy(mFileSystem);
+    int ret;
+    ret = mFsShell.run("cp", "/testWildCards", "/testWildCards/foobar4");
+    Assert.assertEquals(-1, ret);
+    ret = mFsShell.run("cp", "/testWildCards/*", "/testWildCards/foobar4");
+    Assert.assertEquals(-1, ret);
+  }
+
+  private boolean equals(AlluxioURI file1, AlluxioURI file2) throws Exception {
```
Comment (apc999) [inline tests/src/test/java/alluxio/shell/command/CpCommandTest.java]: might be helpful in utils.

File: tests/src/test/java/alluxio/shell/command/CpCommandTest.java (hunk 15 above)
Comment (apc999) [inline tests/src/test/java/alluxio/shell/command/CpCommandTest.java]: discard this

File: shell/src/main/java/alluxio/shell/command/CpCommand.java [hunk 18]
```diff
@@ -0,0 +1,236 @@
...
+        } catch (IOException e) {
+          errorMessages.add(e.getMessage());
+        }
+      }
+
+      if (errorMessages.size() != 0) {
+        throw new IOException(Joiner.on('\n').join(errorMessages));
+      }
+    } else {
+      copyFile(srcPath, dstPath);
```
Comment (apc999) [inline shell/src/main/java/alluxio/shell/command/CpCommand.java]: NIT (readability): when the else branch is so much shorter than the then branch, we should switch the two cases

**Comment by apc999** [comment]
@jsimsa thanks for the PR. I left a bunch comments.

File: shell/src/main/java/alluxio/shell/command/CpCommand.java (hunk 5 above)
Comment (jsimsa) [inline shell/src/main/java/alluxio/shell/command/CpCommand.java]: addressed both your comments (for all shell commands)

File: shell/src/main/java/alluxio/shell/command/CpCommand.java (hunk 12 above)
Comment (jsimsa) [inline shell/src/main/java/alluxio/shell/command/CpCommand.java]: `dstStatus.isFolder()` is needed below

File: shell/src/main/java/alluxio/shell/command/CpCommand.java (hunk 14 above)
Comment (jsimsa) [inline shell/src/main/java/alluxio/shell/command/CpCommand.java]: after addressing an earlier comment the duplicate code is small enough that I don't think it is worth the boilerplate to factor this out

File: shell/src/main/java/alluxio/shell/command/CpCommand.java (hunk 18 above)
Comment (jsimsa) [inline shell/src/main/java/alluxio/shell/command/CpCommand.java]: done

File: shell/src/main/java/alluxio/shell/command/CpCommand.java (hunk 11 above)
Comment (jsimsa) [inline shell/src/main/java/alluxio/shell/command/CpCommand.java]: following the precedent set by CopyToLocalCommand, which does not cache the file copied to the local machine either

File: shell/src/main/java/alluxio/shell/command/CpCommand.java (hunk 1 above)
Comment (jsimsa) [inline shell/src/main/java/alluxio/shell/command/CpCommand.java]: the current style follows the behavior of `CopyFromLocal` / `CopyToLocal` ... are you suggesting I add -R options there as well?

File: shell/src/main/java/alluxio/shell/command/CpCommand.java (hunk 10 above)
Comment (jsimsa) [inline shell/src/main/java/alluxio/shell/command/CpCommand.java]: I think respecting the write type of the source (as opposed to respecting the default write type from the configuration) makes more sense. Otherwise, if you want to copy a large set of files with different write types, you cannot do this using a single command.

File: shell/src/main/java/alluxio/shell/command/CpCommand.java (hunk 13 above)
Comment (jsimsa) [inline shell/src/main/java/alluxio/shell/command/CpCommand.java]: in this code path (`copyWildcard` method) we are copying multiple files onto a directory

File: tests/src/test/java/alluxio/shell/command/CpCommandTest.java (hunk 7 above)
Comment (jsimsa) [inline tests/src/test/java/alluxio/shell/command/CpCommandTest.java]: that is the nature of wildcards ... after the wildcard expansion the behavior will be equivalent to

`cp src1 src2 src3 ... dst` which will not preserve the directory structure either

File: tests/src/test/java/alluxio/shell/command/CpCommandTest.java (hunk 8 above)
Comment (jsimsa) [inline tests/src/test/java/alluxio/shell/command/CpCommandTest.java]: it will create a copy of /testWildCards (before the operation) under /testWildCards/testWildCards

File: tests/src/test/java/alluxio/shell/command/CpCommandTest.java (hunk 9 above)
Comment (jsimsa) [inline tests/src/test/java/alluxio/shell/command/CpCommandTest.java]: done

File: tests/src/test/java/alluxio/shell/command/CpCommandTest.java (hunk 17 above)
Comment (jsimsa) [inline tests/src/test/java/alluxio/shell/command/CpCommandTest.java]: the logic here is specific to the copy command (e.g. comparing the persistence state)

File: tests/src/test/java/alluxio/shell/command/CpCommandTest.java (hunk 8 above)
Comment (jsimsa) [inline tests/src/test/java/alluxio/shell/command/CpCommandTest.java]: I was wrong ... it originally failed ... I will fix this

File: tests/src/test/java/alluxio/shell/command/CpCommandTest.java (hunk 16 above)
Comment (jsimsa) [inline tests/src/test/java/alluxio/shell/command/CpCommandTest.java]: (1) is tested for in copyInvalidArgsTest below
(2) added a test for that

File: tests/src/test/java/alluxio/shell/command/CpCommandTest.java (hunk 8 above)
Comment (jsimsa) [inline tests/src/test/java/alluxio/shell/command/CpCommandTest.java]: fixed

**Comment by jsimsa** [comment]
@apc999 @ifcharming @uronce-cc @aaudiber addressed your comments, PTAL

File: shell/src/main/java/alluxio/shell/command/CpCommand.java (hunk 10 above)
Comment (aaudiber) [inline shell/src/main/java/alluxio/shell/command/CpCommand.java]: When writing the cross-mount move job we decided that it makes more sense to respect the write type configuration over the existing state, so if we want to be consistent we should do the same here

Comment (jsimsa) [inline shell/src/main/java/alluxio/shell/command/CpCommand.java]: OOC, what was the reason behind that decision?

Comment (aaudiber) [inline shell/src/main/java/alluxio/shell/command/CpCommand.java]: 1. All other times when we write data we use the configured write type. 
2. Using the configured write type gives the user some control over how the data is written.

File: tests/src/test/java/alluxio/shell/command/CpCommandTest.java (hunk 8 above)
Comment (uronce-cc) [inline tests/src/test/java/alluxio/shell/command/CpCommandTest.java]: it would be great to add a test here for this case specifically.

File: tests/src/test/java/alluxio/shell/command/CpCommandTest.java (hunk 7 above)
Comment (uronce-cc) [inline tests/src/test/java/alluxio/shell/command/CpCommandTest.java]: thanks for explanation.

File: shell/src/main/java/alluxio/shell/command/CpCommand.java (hunk 10 above)
Comment (jsimsa) [inline shell/src/main/java/alluxio/shell/command/CpCommand.java]: ok, I removed the logic that tries to preserve the cached / persisted state for copied files

File: tests/src/test/java/alluxio/shell/command/CpCommandTest.java (hunk 8 above)
Comment (jsimsa) [inline tests/src/test/java/alluxio/shell/command/CpCommandTest.java]: it is now being tested

File: shell/src/main/java/alluxio/shell/command/CpCommand.java (hunk 1 above)
Comment (ifcharming) [inline shell/src/main/java/alluxio/shell/command/CpCommand.java]: I'm fine with either way. 
Currently `mv` `copyToLocal` `copyFromLocal` does not use `-R` for directories.
Only `ls`, `rm`, `chown` `chgrp` `chmod` require `-R` option.

**Comment by ifcharming** [comment]
LGTM

File: shell/src/main/java/alluxio/shell/command/CpCommand.java (hunk 1 above)
Comment (apc999) [inline shell/src/main/java/alluxio/shell/command/CpCommand.java]: I prefer `cp` to be consistent with `ls` `rm` and `chown` those linux-cmd inspired ones as `cp` also exists for Linux.
For `copyToLocal` and `copyFromLocal`, they are more or less inspired from hdfs shell commands, and people may have less assumption

File: shell/src/main/java/alluxio/shell/command/CpCommand.java (hunk 12 above)
Comment (apc999) [inline shell/src/main/java/alluxio/shell/command/CpCommand.java]: ack

File: shell/src/main/java/alluxio/shell/command/CpCommand.java (hunk 14 above)
Comment (apc999) [inline shell/src/main/java/alluxio/shell/command/CpCommand.java]: ack

File: tests/src/test/java/alluxio/shell/command/CpCommandTest.java (hunk 16 above)
Comment (apc999) [inline tests/src/test/java/alluxio/shell/command/CpCommandTest.java]: SG

**Comment by apc999** [comment]
@jsimsa I put my comment on having "-R" to make `cp` consistent with `mv` and `ls` and etc. Otherwise, I am fine with this PR. Let me know if you feel strong about not having `-R`, and we could discuss offline.

**Comment by uronce-cc** [comment]
LGTM

**Comment by aaudiber** [comment]
LGTM

**Comment by jsimsa** [comment]
@apc999 I added support for -R and more tests (and also did some additional test cleanup)

File: shell/src/main/java/alluxio/shell/command/CpCommand.java (hunk 1 above)
Comment (jsimsa) [inline shell/src/main/java/alluxio/shell/command/CpCommand.java]: done

File: core/common/src/main/java/alluxio/exception/ExceptionMessage.java [hunk 19]
```diff
@@ -128,9 +128,9 @@
       + "the client state due to a timeout."),
 
   // shell
-  DESTINATION_FILE_CANNOT_EXIST_WITH_WILDCARD_SOURCE(
-      "The destination cannot be an existent file when the src is a directory or contains"
-            +  " wildcards."),
+  DESTINATION_CANNOT_BE_FILE(
+      "The destination cannot be a existing file when the source is a directory or a list of "
```
Comment (yupeng9) [inline core/common/src/main/java/alluxio/exception/ExceptionMessage.java]: a -> an

File: shell/src/main/java/alluxio/shell/command/CopyFromLocalCommand.java [hunk 20]
```diff
@@ -62,7 +63,7 @@ protected int getNumOfArgs() {
   }
 
   @Override
-  public void run(CommandLine cl) throws IOException {
+  public void run(CommandLine cl) throws AlluxioException, IOException {
```
Comment (yupeng9) [inline shell/src/main/java/alluxio/shell/command/CopyFromLocalCommand.java]: With the addition of `cp`, this command looks more like `scp` to me.

File: shell/src/main/java/alluxio/shell/command/CpCommand.java [hunk 21]
```diff
@@ -0,0 +1,228 @@
...
+  }
+
+  @Override
+  public String getUsage() {
+    return "cp [-R] <src> <dst>";
+  }
+
+  @Override
+  public String getDescription() {
+    return "Copies a file or a directory in the Alluxio filesystem."
```
Comment (yupeng9) [inline shell/src/main/java/alluxio/shell/command/CpCommand.java]: Does this command support cross-mount copy? If so, are there any caveats?

Comment (jsimsa) [inline shell/src/main/java/alluxio/shell/command/CpCommand.java]: it does ... the caveat is that it is inefficient because it copies everything sequentially

File: core/common/src/main/java/alluxio/exception/ExceptionMessage.java (hunk 19 above)
Comment (jsimsa) [inline core/common/src/main/java/alluxio/exception/ExceptionMessage.java]: done

File: shell/src/main/java/alluxio/shell/command/CopyFromLocalCommand.java (hunk 20 above)
Comment (jsimsa) [inline shell/src/main/java/alluxio/shell/command/CopyFromLocalCommand.java]: hm, I think you make a good point that we could consolidate `copyFromLocal`, `copyToLocal` and `cp` into a single command, which uses `file://` to identify local paths (and non-schemed paths are assumed to be alluxio paths).

I would prefer to do that in a separate PR though

**Comment by jsimsa** [comment]
@yupeng9 thanks for your review ... addressed your comment

**Comment by apc999** [comment]
LGTM thanks for the new command!

File: shell/src/main/java/alluxio/shell/command/CopyFromLocalCommand.java (hunk 20 above)
Comment (yupeng9) [inline shell/src/main/java/alluxio/shell/command/CopyFromLocalCommand.java]: Sounds good. Tracked at ALLUXIO-2024

File: shell/src/main/java/alluxio/shell/command/CpCommand.java (hunk 21 above)
Comment (yupeng9) [inline shell/src/main/java/alluxio/shell/command/CpCommand.java]: sounds good.

**Comment by yupeng9** [comment]
LGTM
